(Fix) Refactor Safe information retrieval

fernandomg opened the pull request:
Aims to close #2048 

We're are having this issue where the Apps or Collectibles sections disappear.

One of the possible reasons is that we're managing the update of Safe's information from many sources. Two slightly different functions for the same goal (update safe's info), as well as loading data from the `transaction-service` and from RPC.

Here a refactor was made for the `fetchSafe` action. `checkAndUpdateSafe` action was deleted. And `buildSafe` was refactored too. Now `fetchSafe` and `buildSafe` ~share quite similar logic~ are clearly differentiated.

On the technical side, instead of using `Promise.all`, we used `.allSettled` as it allows for better flexibility while we're retrieving Safe's data. Data is requested from `client-gateway` instead of a mix of `transaction-service` and `RPC calls`.

This refactor aims to leave the functionality untouched (`fetchSafe` and `buildSafe` both do the exact same tasks they were doing until now).

Things I noticed that we should work on from here:
1. load safe flow, can be improved
2. prevent unnecessary calls to `fetchSafe` (especially when loading from URL)
3. continue deleting legacy code.

## How to test
Given that this is a refactor on how we recover and load the safe information, the test should be as thorough as possible.

What I certainly recommend checking is:
* Every 'load safe' flow, should behave as in production.
* Every 'create safe' flow, should behave as in production (with a slight enhancement, see: https://github.com/gnosis/safe-react/pull/2122#issuecomment-821836310)
* No more 'app' or 'collectibles' sections disappearing

rmeissner commented on src/logic/safe/utils/modules.ts:
```diff
@@ -65,26 +60,28 @@ export const getModules = async (safeInfo: SafeInfo | void): Promise<Array<Modul
 
+  const safeInfoModules = safeInfo.modules.map(({ value }) => value)
+
   if (semverLessThan(safeInfo.version, '1.1.1')) {
```
Is this if else logic required? The backend should always return all modules.

Maybe check with @Uxio0 on that and also make sure that the backend always returns the correct order (as order is important for this logic here)

fernandomg replied:
If I'm not wrong, `gnosis-py` is only using `getModules` https://github.com/gnosis/gnosis-py/blob/master/gnosis/safe/safe.py#L564

rmeissner replied:
Fair, I would still not add this logic here ... then better open an issue on the gnosis py to add support for this.

fernandomg replied:
Yes, agree. This was here until the change in gnosis-py was made.

I'll open the issue and add a TODO in this block pointing to it.

Thanks!

Uxio0 replied:
Unless `10` modules are returned so we don't know if there're more, and then we fallback to the paginated version https://github.com/gnosis/gnosis-py/blob/master/gnosis/safe/safe.py#L571 😉

fernandomg replied:
:scream: ahhh... I'll close the issue I just open :grimacing:

fernandomg commented:
Just added the last modification to this PR.

It's how the owner's are retrieved/stored in memory on `fetchSafe`

/cc @nicosampler, can you re-check please?

dasanra reviewed (changes requested):
Good start, but I found some small things ;)

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
-  const featuresEnabled = enabledFeatures(currentVersion)
-  const modules = await getModules(safeInfo)
-  const spendingLimits = safeInfo ? await getSpendingLimits(safeInfo.modules, safeAddress) : null
+  return safeInfo
```
only the `name` will be returned if we managed to retrieve the information from the backend.

fernandomg replied:
I realized that my previous comment was wrong.

As you said, that was the case. Now if there's no `remoteSafeInfo`, we would not call this function.

katspaugh commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+  const [remoteSafeInfoSettled, localSafeInfoSettled, safeBalancesSettled] = await Promise.allSettled([
     getSafeInfo(safeAddress),
     getLocalSafe(safeAddress),
+    fetchTokenCurrenciesBalances({ safeAddress, selectedCurrency: 'UDS' }),
```
We should probably make it an enum or constant. It's used in a few other places.

fernandomg replied:
Good news, this is no longer used :tada:

katspaugh commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
-      )
-    }
-  })
+  return { ...safeInfo, ...remoteSafeInfo, ...localSafeInfo, owners } as SafeRecordProps
```
Why is `localSafeInfo` going after the `remoteSafeInfo`?
Shouldn't the backend be the source of truth for the initial load?

dasanra replied:
Maybe the approach of changing the spread order can be more interesting. I was also having troubles while checking the local values info function. I don't think that avoid adding the values to localSafeInfo save us load effort and could lead to a more simple code. :point_up:

fernandomg replied:
along with other changes, the spread was modified, which simplified how local safe info is recovered, which lead to simply delete the `extractLocalSafeInfo` helper function.

Actually, the only place where `localSafeInfo` is used, is in `buildSafe` function (we may be able to load the information from the redux store, but that's something to work on the "load safe flow improvement")

mikheevm commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+
+    remoteOwners = remoteSafeInfo.owners.map(({ value }) => {
+      const localOwner = localOwners?.find(({ address }) => sameAddress(address, value))
+      const name = localOwner?.name ?? 'UNKNOWN'
```
@katspaugh In my understanding, it's a mix of privacy concerns and technical complexity. If we store them on the backend, they have to be either public or private. The public is not ideal,  it could expose someone's identity. If you make it private, you need to authorize the user to view them somehow which could be complex. Maybe you can raise this question with Richard

katspaugh replied:
Gotcha, thanks!
Still need to wrap my head around the everything-is-public concept.

katspaugh commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+
+  if (remoteSafeInfoSettled.status === 'fulfilled') {
+    const remoteSafeInfo = remoteSafeInfoSettled.value
+    const safeInfoModules = remoteSafeInfo.modules.map(({ value }) => value)
```
Everything related to modules from here to line 61 could be skipped if the module list is empty.

fernandomg replied:
done! :male_detective:

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+): Promise<Partial<SafeRecordProps>> => {
+  const safeInfo: Partial<SafeRecordProps> = {}
+
+  if (remoteSafeInfoSettled.status === 'fulfilled') {
```
done, thanks for the suggestion.

I ended up letting the `allSettled` name to the helper function that handles the promises :man_shrugging:

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
-      address: safeAddress,
-      methods: safeParams,
-    }),
+export const buildSafe = async (safeAddress: string, safeName: string): Promise<SafeRecordProps> => {
```
after the refactor, there's almost no duplicate code. Each function has its unique differentiated logic.

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+  let localOwners
+  let remoteOwners
+
+  if (localSafeInfoSettled.status === 'fulfilled' && localSafeInfoSettled.value !== undefined) {
```
this is no longer there, code was greatly simplified with the `allSettled` helper function, and with the removal of `extractLocalSafeInfo` fuction.

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
+
+    remoteOwners = remoteSafeInfo.owners.map(({ value }) => {
+      const localOwner = localOwners?.find(({ address }) => sameAddress(address, value))
+      const name = localOwner?.name ?? 'UNKNOWN'
```
now `UNKOWN` is the default value for the owner's record creation (`makeOwner`).

fernandomg commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,187 @@
 
-  dispatch(updateSafe(updatedSafe))
+  // remote (client-gateway)
+  // TODO: REVIEW: do we really need to load ethBalance?
```
removed this, and everything is working fine. All balance-related information is being loaded by `fetchSafeTokens`.

fernandomg commented:
Sorry @nicosampler, one more time. A rather interesting refactoring led to a more simplified overall logic.

katspaugh commented on src/logic/safe/store/actions/fetchSafe.ts:
```diff
@@ -1,178 +1,140 @@
-const buildOwnersFrom = (safeOwners: string[], localSafe?: SafeRecordProps): List<SafeOwner> => {
-  const ownersList = safeOwners.map((ownerAddress) => {
-    const convertedAdd = checksumAddress(ownerAddress)
+  if (safeInfoModules) {
```
This will always be true. I think we need to check the length of the array.

fernandomg replied:
:grimacing:

katspaugh reviewed (changes requested):
Looks great! Love the extra docs!
Just one small comment.

fernandomg commented:
> Looks great! Love the extra docs!
> Just one small comment + there are some eslint warnings.

@katspaugh, fixed the modules check.

In regard to eslint warnings, for middlewares, I'll wait to do it in a separate PR as it's something that needs to be done for all middlewares. And `updateSafeName` is deleted in #2123.

dasanra commented on src/logic/safe/utils/allSettled.ts:
```diff
@@ -0,0 +1,22 @@
+  const data = new Array(args.length).fill(null)
+  const promises = await Promise.allSettled(args)
+
+  for (let i = 0; i < promises.length; i++) {
```
I'm more for the `promises.forEach` syntax ;)

fernandomg replied:
changed

dasanra reviewed (approved):
This should be ready for a QA check

fernandomg commented:
As per @katspaugh, I added tests for the `allSettled` function. [`4fa7ce7` (#2122)](https://github.com/gnosis/safe-react/pull/2122/commits/4fa7ce78149e8cce82f6682abde62e8f0cc790f5)

francovenica commented:
1 - I created a safe. The assets table took a while to load, while it was loading the Apps button and the Collectibles button at the top was missing. Once loaded the buttons show up just fine
![image.png](https://images.zenhubusercontent.com/5e14b8e3edee21d6f9aa6c8a/bbccfe82-1111-4b34-a8f5-305e10392756)

All the following issues are for both, the "load safe" and "create safe" processes: 
NOTE: If all the following are not actual issues but something it had to be done to fix the bug please make a comment.

2 - I created a safe and set 3 owners, I gave names to all 3 owners, but none were saved, they show up as UNKNOWN in the settings.

After this I edited the names in the settings and I had no issues with that; the names changed and they were added automatically in the address book. Same editing in the address book, the name also changes in the settings. So the issue is only in the creation process

3 - Related to the last issue. The name of the safe is saved properly (you can see it fin in the settings), but is not added to the address book automatically. 
If you edit the name in the settings the safe is added to the address book just fine, so again the issue is in the safe creation process

4 - Deleting an owner from the address book will reset his name in the settings as "UNKNOWN", that should not happen. the name should persist in the settings owner table
Note: if you delete the name in the address book and go straight to the settings > owners table you might see that the name still persist, but in the next render of the page (the every 15 secs render) the name will change to UNKNOWN

francovenica reviewed (changes requested):
Check previous my previous comment

katspaugh commented:
It would be nice to emulate the `UNKNOWN` owner case as a unit test before fixing it.

fernandomg commented:
I added a proposal to fix the first issue that @francovenica reported (the issue with no App, nor Collectibles being displayed)

Before changes in this PR we always relied on `localStorage` when we fetched safe information. This helped with the "Create Safe" flow, as we optimistically store the information locally and waited for the services to finally provide their data.

If we keep relying on `localStorage` it will over-complicate the logic inside `fetchSafe` or even in the `UPDATE_SAFE` reducer, as we need to be aware of what flow we're in.

So the solution goes around _waiting_ for the client-gateway to finally provide information about the safe, and then navigate the user to the `balances` screen.

https://github.com/gnosis/safe-react/blob/ba51a33b4f9e156f5b8eee6596e6aa894ff9fc6c/src/routes/open/container/Open.tsx#L194-L203

One point against all this is that the user will have to wait a little while for his newly created Safe.

@dasanra, @katspaugh, @nicosampler WDYT?

Note: this is the solution I came with after finding the source of the issue. I'm not saying that if the strategy is correct it will be the final code :grin: 

How it looks:
![waiting-for-client-gateway](https://user-images.githubusercontent.com/3315606/115117351-743b5400-9f74-11eb-876e-8440c7611066.gif)


In regard to the addressBook-related issues, I proposed to work on them on a separate issue.

katspaugh commented:
Makes sense to me.

I would add a circuit breaker in that while loop though. It can potentially spam the API forever.
E.g. if the gateway responds with a 50x error, or after 100 unsuccessful attempts.

katspaugh commented:
We could also increment the delay in each iteration like so:
```
var delayMs = 1000;
for (let i = 0; i < 20; i++) {
  delayMs *= 1.1;
  console.log(Math.round(delayMs));
}
1100
1210
1331
1464
1611
1772
1949
2144
2358
2594
2853
3138
3452
3797
4177
4595
5054
5560
6116
6727
```

mikheevm commented:
> We could also increment the delay in each iteration like so:
> 
> ```
> var delayMs = 1000;
> for (let i = 0; i < 20; i++) {
>   delayMs *= 1.1;
>   console.log(Math.round(delayMs));
> }
> 1100
> 1210
> 1331
> 1464
> 1611
> 1772
> 1949
> 2144
> 2358
> 2594
> 2853
> 3138
> 3452
> 3797
> 4177
> 4595
> 5054
> 5560
> 6116
> 6727
> ```

I think we're using exponential backoff somewhere in the app, the logic can be reused

katspaugh commented:
@fernandomg please consider these comments just as ideas for the next iteration. I think the code is already good for now.

fernandomg commented:
@katspaugh, sorry for the really slow/late response, but I was having issues with my local environment.
<br>

> 1 - I created a safe. The assets table took a while to load, while it was loading the Apps button and the Collectibles button at the top was missing. Once loaded the buttons (Apps and Collectibles) show up just fine

@francovenica, please, re-check this issue.

**do not merge this to `development`**. It should be merged after the code freeze.

katspaugh commented on .nvmrc:
```diff
@@ -1 +1 @@
-10
+14
```
Good move! Not sure if it's going to affect the deployment.

fernandomg replied:
no, just local dev. We have the version fixed to 14 in the CI (which happens to be 14 too).

katspaugh commented on src/routes/open/container/Open.tsx:
```diff
@@ -192,7 +191,18 @@ const Open = (): React.ReactElement => {
+    await sleep(5000)
+
+    await backOff(() => getSafeInfo(safeAddress), {
+      startingDelay: 750,
```
Awesome! 👏 
Will the 10 default attempts be enough?

fernandomg replied:
From what I tested, yes. But it will let the button in an infinite "loading" state if we reach the maximum of attempts 🤔

fernandomg replied:
hm..., it will redirect to the `/balances` endpoint anyway. I believe that for the current state of affairs it is a, so far, good solution.

katspaugh commented on src/routes/opening/components/Footer.tsx:
```diff
@@ -75,7 +75,13 @@ export const ContinueFooter = ({
-      Get started
+      {continueButtonDisabled ? (
+        <>
+          <Loader size="xs" color="secondaryLight" /> Loading your safe
```
We need a small gap between the loader and the text. Rn they are right next to each other.

katspaugh commented:
@francovenica do you think the issues with owner names and address book entries were introduced in this PR?

I see similar problems on the development branch.

francovenica commented:
@katspaugh At least they are not happening in prod right now, so Idk if it was introduced by this ticket, but it was during this sprint. I think no other ticket dealt with the storage of owners names in this way.

francovenica commented:
@fernandomg As we discussed I'll report here what is still broken. If you think these should be solved in the 2182 we can put them there. Probably are all related

1 - Loading safe workflow. With a clean storage, when the safe has done loading the address book remains empty. It should have the safe with its name and the owners who were given names during the process

2 - The owners have "Unknown" name in the settings after loading the safe, even when names where given to them.

3 - Removing the owner entry from the address book keeps defaulting his name to "UNKNOWN" in the settings owner's list. This time it takes 15 secs to update (So if you delete the entry and go straight to settings, you might still see his name, but after 15 secs top his name will reset to Unknown

4 - The address book is emptied when you move out from the safe you are and comeback, for example
Go a safe you have in the side bar and then to the the address book
Add one or 2 entries
Go to the safe creation, just the first step, do nothing there
Go back to the safe > address book
The addressbook will be empty

francovenica commented:
Ivan created this PR to fix the 4 issues commented above. 
https://github.com/gnosis/safe-react/pull/2190

They are fixed there

katspaugh commented:
I've merged that PR.
@fernandomg please update the branch and the address book issues should be gone.

katspaugh commented:
Since we now have time until 3.6.0, I suggest to cover `fetchSafe` with a couple tests.
* Mock the remote and local safe info
* Test the case with the local storage populated
* Test the case without an empty local storage

fernandomg commented:
> Since we now have time until 3.6.0, I suggest to cover `fetchSafe` with a couple tests.
> 
> * Mock the remote and local safe info
> * Test the case with the local storage populated
> * Test the case without an empty local storage

@katspaugh, well, I think most of the cases are covered. Especially those functions that are in charge of merging _remote_ and _local_ information.

Things that were modified due to tests implementation:
* `extractRemoteSafeInfo` was moved into `utils.ts`
* `extractSafeOwners` was moved into `utils.ts` and renamed to `buildSafeOwners` (as it's building a definitive list based on remote and local info)
* `extractRemoteSafeInfo` arguments were reduced to `1` argument. `address` was redundant, as that info was provided by `remoteSafeInfo`.
* in `extractRemoteSafeInfo` the modules/spendingLimits data recovery was simplified
  * no need to call `getModules` as that info was provided by `remoteSafeInfo`
  * `getModules` was deleted, as it was useful before having the information retrieved and provided by the transactions-service
* the order of the information spread by `buildSafe` was modified [`bccc590` (#2122)](https://github.com/gnosis/safe-react/pull/2122/commits/bccc59096b2d3fe2db5ee10580422d93daa5a384#diff-12f45c5c5b8715f2211b48943692986d74cdb6544fda52eb55ff5d3451262266R42)
* added `redux-mock-store` into `devDependencies` to properly test redux actions payload.

katspaugh reviewed (approved):
Looks amazing! Thanks a lot!

francovenica commented:
I didn't had any issues with the app button or collectibles link after roaming around the site for while, so that's fine

Regarding the issues with the addressbook:
- Issues 1, 2 and 4 were fixed

- Issue 3 is still present
- A new issue: 5 - When a safe is loaded / created is not being saved in the address book. In prod the safe is being added as an entry after loading/creating it.

Let me know if this issues about the address book will be addressed here or if we should open a new PR like we did before.

fernandomg commented:
> * A new issue: 5 - When a safe is loaded / created is not being saved in the address book. In prod the safe is being added as an entry after loading/creating it.

@francovenica, this is sort of related to AddressBook. But I added a workaround for it, as it's odd to have the inconsistency between Address Book and Settings.

francovenica commented:
5 - This is fixed

Given the talks we were having, I'll assume the issue 3 will stay that way.
